### DISCUSSION TITLE
How to setup DCUTR properly?

### DISCUSSION THREAD

**Discussion by shamil-gadelshin** [libp2p/rust-libp2p#5291] [General]
We're researching an option to add `DCUTR` to our network and have a couple of questions about the protocol. 

1. Do we need `port_reuse` with `DCUTR`? We found this discussion:https://github.com/libp2p/specs/issues/389
And it seems that `port_reuse` setting is not mandatory.  If it is not then how src and dest peers will initiate two connections with the same ports and ip addresses (port will be random on both sides)? 

2. If `port_reuse` is necessary then how autonat will work? 
One of the common ways to use autonat is to request for a probe from the currently connected peers. But the new connection will fail because of the duplicate tuple of (source IP, source port, destination IP, destination port) when `port_reuse` enabled. Am I missing something here? 
This comment from the discussion looks like a workaround: https://github.com/libp2p/specs/issues/389#issuecomment-1433814190
However, is there a way to toggle the port reuse flag in the configured swarm?

3. Are there any projects that use rust version of `DCUTR`? 
We found this project: https://github.com/n0-computer/beetle/blob/main/iroh-p2p/src/behaviour.rs
But it is archived now and it's unknown whether its DCUTR worked in the first place (no port reuse though).

4. How to setup DCUTR properly? 
My current understanding of the DCUTR contradicts itself: 
according to the specification - peers need to simultaneously connect to each other having the identical tuple (source IP, source port, destination IP, destination port) to hole punch which is achieved using port reuse but autonat won't work in this setup because it won't create a second connection with already connected peer for a probe because of the existing tuple (src IP, src port, dest IP, dest port). I skip the relayer part in this description for simplicity. Please, point to an error in my reasoning.

**Comment by dhuseby** [comment]
@DougAnderson444 do you have any experience with the DCUTR implementation in rust-libp2p?

**Reply by DougAnderson444** [reply]
Unfortunately I do not :/

**Comment by drHuangMHT** [comment]
DCUtR is handled automatically when two peers who support it connects to a common relay. However I haven't yet observed a case where the direct connection upgrade succeeded.
